src/collector.rs

```diff
@@ -1,103 +1,175 @@
+            start_time_cycles: self.start_time_cycles,
+            elapsed_ns,
+            cycles_per_second: minstant::cycles_per_second(),
+            spans: span_set.spans,
+            properties: span_set.properties,
         }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TraceResult {
+    pub trace_id: u64,
+
+    pub start_time_ns: u64,
+    pub start_time_cycles: u64,
+    pub elapsed_ns: u64,
+
+    /// For conversion of cycles -> ns
+    pub cycles_per_second: u64,
+
+    /// Span collection
+    pub spans: Vec<Span>,
+
+    /// Properties
+    pub properties: Properties,
+}
+
+/// Properties can used to attach some information about tracing context
+/// to current span, e.g. host of the request, CPU usage.
+///
+/// Usage:
+/// ```
+/// # let event_id = 1u32;
+/// let _guard = minitrace::new_span(event_id);
+/// minitrace::new_property(b"host:127.0.0.1");
+/// minitrace::new_property(b"cpu_usage:42%");
+/// ```
+///
+/// Every property will relate to a span. Logically properties are a sequence
+/// of (span id, property) pairs:
+/// ```text
+/// span id -> property
+/// 10      -> b"123"
+/// 10      -> b"!@$#$%"
+/// 12      -> b"abcd"
+/// 14      -> b"xyz"
+/// ```
+///
+/// and will be stored into `Properties` struct as:
+/// ```text
+/// span_ids: [10, 10, 12, 14]
+/// property_lens: [3, 6, 4, 3]
+/// payload: b"123!@$#$%abcdxyz"
+/// ```
+#[derive(Debug, Clone)]
+pub struct Properties {
+    pub span_ids: Vec<u32>,
+    pub property_lens: Vec<u64>,
+    pub payload: Vec<u8>,
+}
+
+#[derive(Debug, Clone)]
+pub struct SpanSet {
+    /// Span collection
+    pub spans: Vec<Span>,
 
-        let mut spans = Vec::with_capacity(spans_len);
-        let mut span_ids = Vec::with_capacity(span_ids_len);
-        let mut property_lens = Vec::with_capacity(property_lens_len);
-        let mut payload = Vec::with_capacity(payload_len);
+    /// Property collection
+    pub properties: Properties,
+}
 
-        for span_set in &span_sets {
-            spans.extend_from_slice(&span_set.spans);
-            span_ids.extend_from_slice(&span_set.properties.span_ids);
-            property_lens.extend_from_slice(&span_set.properties.property_lens);
-            payload.extend_from_slice(&span_set.properties.payload);
+impl SpanSet {
+    pub fn new() -> Self {
+        SpanSet {
+            spans: Vec::new(),
+            properties: Properties {
+                span_ids: Vec::new(),
+                property_lens: Vec::new(),
+                payload: Vec::new(),
+            },
         }
+    }
 
-        crate::TraceDetails {
-            start_time_ns: self.inner.start_time_ns,
-            elapsed_ns: crate::time::real_time_ns().saturating_sub(self.inner.start_time_ns),
-            cycles_per_second: minstant::cycles_per_second(),
-            spans,
-            properties: crate::Properties {
-                span_ids,
-                property_lens,
-                payload,
+    pub fn with_capacity() -> Self {
```

Comment:
A more reasonable name?

Reply:
I think `with_capacity` is reasonable, especially in the circumstance that the `SpanSet` has 
 a function `new()`, which returns `SpanSet` with no capacity.